[Outlaw] Initial Outlaw T31 set effect & 10.2 PTR talent changes

weswigham opened the pull request:
Included Outlaw PTR work items:
- [x] T31 set
- [x] Roll the Bones
- [x] Restless Blades
- [x] Between the Eyes
- [x] Swift Slasher
- [x] Deft Maneuvers
- [x] Killing Spree
- [x] Ghostly Strike
- [x] Improved Adrenaline Rush
- [x] Underhanded Upper Hand
- [x] Crackshot
- [x] Sanity check that everything in the updated base class tree is still working as intended with PTR data (MFD?)

weswigham commented on engine/class_modules/sc_rogue.cpp:
```diff
@@ -6588,6 +6697,21 @@ struct adrenaline_rush_t : public buff_t
+    if ( rogue->talent.outlaw.underhanded_upper_hand->ok() && rogue->buffs.blade_flurry->check() )
+    {
+      auto current_duration = rogue->buffs.blade_flurry->remains();
+      rogue->buffs.blade_flurry->extend_duration( rogue, (current_duration * 100.0) - current_duration );
```
Is this a janky way to "pause" a buff? Absolutely. But it basically matches what's done by the game, which sets a `modRate` of 100 to make them count down slower, while still displaying the same number on these buffs while they're "paused". The "paused" buffs can, in fact, still expire - just takes awhile.

EvanMichaels commented:
Discussed a bit with Solo and Zac in Ravenholdt. I think most of this is good, although I have some concerns about some of it. (Killing Spree, as an example, I think could be done in a bit of a better way.) But this is a pretty large PR to review all at once so I may not be able to review and comment on everything immediately. May be easier to reach out on Discord about this.

weswigham committed:
Support applying buff time modifiers

This only supports applying them by label, and does not currently support
modifying the tick rate or apparent duration of a buff.

weswigham committed:
Implement outlaw 10.2 PTR changes

weswigham committed:
Finish Prey on the Weak -> String like a Bee move and rename

weswigham committed:
Finish reorganizing base class tree to new 10.2 PTR shape

weswigham committed:
Nightstalker applies in Subterfuge (and maybe always should have on live, too)

weswigham committed:
Missed some nightstalker references

weswigham committed:
Finish implementing new MFD - the damage amp is _not_ snapshotted by DoTs (but it does affect them)

weswigham committed:
Move BTE crit calculation into more appropriate function(s) since its now a buff

weswigham committed:
Actually, remove special BTE handling entirely - its correctly handled by generic buff mechanisms, and we were double dipping

weswigham committed:
Blade flurry generating CP should require deft maneuvers

weswigham committed:
Track dreadblades self damage if dps heal tracking is on. For funsies (mostly lets you know if your leech is sufficient to sustain your db usage, since soothing darkness/recup heals arent really implemented).

weswigham committed:
Fix BTE buff refresh behavior

weswigham committed:
Add secondary action damage modifiers for crackshot and fan the hammer

weswigham committed:
Remove MFD and Dreadblades

weswigham committed:
Implement grand melee changes

weswigham committed:
Move airborne irritant, remove incorrect comment

weswigham committed:
Move around talent initializers to match current tree again

weswigham committed:
Ahaha, dreadblades is gone, but now theres two Ghostly Strikes

weswigham committed:
Rename functions and variables to clarify only duration is currently implemented

weswigham committed:
Update for new base buff

weswigham committed:
Fix cached buff reset between iterations

Also, just inline this function, it's simpler inline anyway.

weswigham committed:
Move prepull buff reset into init/reset so as to not rely on a prepull rtb to reset state

weswigham committed:
Put PS damage reduction behind the bugs flag - hopefully this gets fixed

weswigham committed:
Split static and dynamic multiplier parts, reset dynamic on buff reset, per reccomentdation

weswigham committed:
Support renamed buff function post-rebase

weswigham committed:
Add sanity check on input to function

weswigham committed:
Renamings round 2

weswigham committed:
Add precision shot, audit no implementation talents while at it

weswigham committed:
This doesnt need to be public anyway

weswigham committed:
Fix swift slasher calculation - composite_melee_haste is a multiplier, not a %
